### PR TITLE
Merge query and read feed

### LINKED ISSUE #459: Merge Read Feed and Query
Right now read feed and query are two separate operations. This is rather confusing concept since read feed is just a "select *" query. Instead of adding additional APIs for this operation the SDK implementation should handle the selection. 

Convert all Get*Iterators -> Get*QueryIterators
QueryText == null will be converted to a read feed operation
Non-Item operations will be a stateless query that will go to the gateway to get the results.

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This merges query and read feed. Added a new stateless query that handle all the logic for non-partitioned resources.

Scripts still needs to be updated.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

closes #459

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -407,15 +409,40 @@ public virtual Container GetContainer(string databaseId, string containerId)
...
-                nextDelegate: this.DatabaseFeedRequestExecutorAsync);
+                requestOptions);
+
+            return new FeedStatelessIteratorCore<DatabaseProperties>(
+                databaseStreamIterator,
+                this.ClientContext.ResponseFactory.CreateResultSetQueryResponse<DatabaseProperties>);
+        }
+
+        /// <summary>
+        /// Gets an iterator to go through all the containers for the database
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs]: Please remove container

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 2]
```diff
@@ -407,15 +409,40 @@ public virtual Container GetContainer(string databaseId, string containerId)
...
+        /// </summary>
+        /// <param name="queryDefinition">The cosmos SQL query definition.</param>
+        /// <param name="continuationToken">The continuation token in the Azure Cosmos DB service.</param>
+        /// <param name="requestOptions">(Optional) The options for the container request <see cref="QueryRequestOptions"/></param>
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs]: container  -> database ?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 3]
```diff
@@ -390,8 +391,9 @@ public virtual Container GetContainer(string databaseId, string containerId)
         /// <summary>
         /// Gets an iterator to go through all the databases for the account
         /// </summary>
-        /// <param name="maxItemCount">The max item count to return as part of the query</param>
+        /// <param name="queryDefinition">The cosmos SQL query definition.</param>
         /// <param name="continuationToken">The continuation token in the Azure Cosmos DB service.</param>
+        /// <param name="requestOptions">(Optional) The options for the item query request <see cref="QueryRequestOptions"/></param>
         /// <example>
         /// Get an iterator for all the database under the cosmos account
         /// <code language="c#">
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs]: Want to add example 2 with non null queryDefinition ?

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 4]
```diff
@@ -149,7 +149,7 @@ internal override void PopulateRequestOptions(RequestMessage request)
         {
             request.Headers.Add(HttpConstants.HttpHeaders.ContentType, MediaTypes.QueryJson);
             request.Headers.Add(HttpConstants.HttpHeaders.IsQuery, bool.TrueString);
-            request.Headers.Add(HttpConstants.HttpHeaders.EnableCrossPartitionQuery, bool.TrueString);
+            request.Headers.Add(HttpConstants.HttpHeaders.EnableCrossPartitionQuery, bool.FalseString);
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: why this is flipped ?

File: Microsoft.Azure.Cosmos/src/Resource/Database/Database.cs [hunk 5]
```diff
@@ -389,16 +389,35 @@ public abstract class Database
...
-        /// <param name="maxItemCount">(Optional) The max item count to return as part of the query</param>
+        /// <param name="queryDefinition">The cosmos SQL query definition.</param>
         /// <param name="continuationToken">(Optional) The continuation token in the Azure Cosmos DB service.</param>
+        /// <param name="requestOptions">(Optional) The options for the item query request <see cref="QueryRequestOptions"/></param>
         /// <example>
         /// Get an iterator for all the containers under the database
         /// <code language="c#">
         /// <![CDATA[
-        /// FeedIterator<ContainerProperties> feedIterator = this.cosmosDatabase.GetContainerIterator();
+        /// FeedIterator<ContainerProperties> feedIterator = this.cosmosDatabase.GetContainerQueryIterator();
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/Resource/Database/Database.cs]: Any example for  not null queryDefinition  in docs?

File: Microsoft.Azure.Cosmos/src/Resource/QueryResponses/QueryDefinition.cs [hunk 6]
```diff
@@ -68,6 +68,15 @@ public QueryDefinition UseParameter(string name, object value)
             return this;
         }
 
+        /// <summary>
+        /// Implicit converter
+        /// </summary>
+        /// <param name="queryText">The query text</param>
+        public static implicit operator QueryDefinition(string queryText)
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/Resource/QueryResponses/QueryDefinition.cs]: Will discuss this , my 2 cents - it is really hard for user to figure out the implicit conversion and I guess taking string was our idea to make it simple for user who just want to use non parameterised query.

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: This is a bug. I sent out the PR to get feedback on the design. I'll fix the comments and add more test after I the design is approved.

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 7]
```diff
@@ -830,59 +766,17 @@ public abstract class Container
...
-        ///         }
-        ///     }
-        /// }
-        /// ]]>
-        /// </code>
-        /// </example>
-        /// <returns>An iterator to go through the items.</returns>
-        public abstract FeedIterator GetItemQueryStreamIterator(
-            string sqlQueryText,
+            QueryDefinition queryDefinition,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: Optional?

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 8]
```diff
@@ -830,59 +766,17 @@ public abstract class Container
...
+            PartitionKey partitionKey = null,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: Any specific reason its promoted now?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: It matches the other item operations, and it prevents having partition key in request options which is shared by database/container operations.

File: Microsoft.Azure.Cosmos/src/Handler/RequestMessage.cs [hunk 9]
```diff
@@ -167,9 +167,17 @@ internal DocumentServiceRequest ToDocumentServiceRequest()
...
-                    serviceRequest = new DocumentServiceRequest(this.OperationType, null, this.ResourceType, this.Content, this.Headers.CosmosMessageHeaders, false, AuthorizationTokenType.PrimaryMasterKey);
+                    serviceRequest = new DocumentServiceRequest(
+                        operationType: this.OperationType,
+                        resourceIdOrFullName: null,
+                        resourceType: this.ResourceType,
+                        body: this.Content,
+                        headers: this.Headers.CosmosMessageHeaders,
+                        isNameBased: false,
+                        authorizationTokenType: AuthorizationTokenType.PrimaryMasterKey);
+                    serviceRequest.UseGatewayMode = true;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/RequestMessage.cs]: Why? Did any tests fail?
Metadata operaitons are also served in Direct mode.

File: Microsoft.Azure.Cosmos/src/Query/ParallelQuery/ItemProducer.cs [hunk 10]
```diff
@@ -322,6 +323,8 @@ public async Task BufferMoreDocumentsAsync(CancellationToken token)
                         QueryRequestOptions.FillMaxItemCount(
                             cosmosRequestMessage,
                             pageSize);
+                        cosmosRequestMessage.Headers.Add(HttpConstants.HttpHeaders.ContentType, MediaTypes.QueryJson);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/ParallelQuery/ItemProducer.cs]: Why this change demand them?

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 11]
```diff
@@ -230,10 +239,12 @@ internal FeedOptions ToFeedOptions()
             RequestMessage request,
             string continuationToken)
         {
-            if (!string.IsNullOrWhiteSpace(continuationToken))
+            if (continuationToken == null)
             {
-                request.Headers.Add(HttpConstants.HttpHeaders.Continuation, continuationToken);
+                continuationToken = string.Empty;
             }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: Is CT always demanded. I.e. why not set only when its != NULL

File: Microsoft.Azure.Cosmos/src/Resource/Conflict/Conflicts.cs [hunk 12]
```diff
@@ -118,8 +121,59 @@ public abstract class Conflicts
         /// ]]>
         /// </code>
         /// </example>
-        public abstract FeedIterator GetConflicttreamIterator(
-            int? maxItemCount = null,
-            string continuationToken = null);
+        public abstract FeedIterator GetConflictQueryStreamIterator(
+            QueryDefinition queryDefinition,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Conflict/Conflicts.cs]: >QueryDefinition [](start = 12, length = 15)

Can QueryDefinition overload wait? String version is generic and will server better. 
Usage of query for meta-data it-self would be lower and we can bring them back on demand. 
Thoughts?

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 13]
```diff
@@ -896,31 +828,35 @@ public abstract class Container
...
         ///     }
         /// }
         /// ]]>
         /// </code>
         /// </example>
         /// <returns>An iterator to go through the items.</returns>
-        public abstract FeedIterator<T> GetItemQueryIterator<T>(
-            QueryDefinition sqlQueryDefinition,
+        public abstract FeedIterator GetItemQueryStreamIterator(
+            string queryText = null,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: > = null [](start = 28, length = 7)

Wil it help in any scenarios? If not why to default it?

File: Microsoft.Azure.Cosmos/src/Query/ParallelQuery/ItemProducer.cs (hunk 10 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/ParallelQuery/ItemProducer.cs]: We are using query request options for both read feed and query. This isn't header isn't required for read feed so I moved it only into the query pipeline.

File: Microsoft.Azure.Cosmos/src/Resource/Conflict/Conflicts.cs (hunk 12 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Conflict/Conflicts.cs]: QueryDefinition is required to prevent SQL injection. Only supporting string is a security issue.

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs (hunk 13 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: This helps prevent ambiguity in the overloads. If I remove the default it will require the user to always parameterize the input to get read feed since both string and queryDefinition support null.

This is current way to get read feed
`container.GetItemQueryStreamIterator()`

Removing default requires this:
`container.GetItemQueryStreamIterator(queryDefintion: null)`
or 
`container.GetItemQueryStreamIterator(queryText: null)`

File: Microsoft.Azure.Cosmos/src/Handler/RequestMessage.cs [hunk 14]
```diff
@@ -167,9 +167,16 @@ internal DocumentServiceRequest ToDocumentServiceRequest()
             if (this.DocumentServiceRequest == null)
             {
                 DocumentServiceRequest serviceRequest;
-                if (this.OperationType == OperationType.ReadFeed && this.ResourceType == ResourceType.Database)
+                if ((this.OperationType == OperationType.ReadFeed) && this.ResourceType == ResourceType.Database)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/RequestMessage.cs]: Unnecessary surrounding ()

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 15]
```diff
@@ -147,9 +149,16 @@ public class QueryRequestOptions : RequestOptions
         /// <param name="request">The <see cref="RequestMessage"/></param>
         internal override void PopulateRequestOptions(RequestMessage request)
         {
-            request.Headers.Add(HttpConstants.HttpHeaders.ContentType, MediaTypes.QueryJson);
-            request.Headers.Add(HttpConstants.HttpHeaders.IsQuery, bool.TrueString);
-            request.Headers.Add(HttpConstants.HttpHeaders.EnableCrossPartitionQuery, bool.TrueString);
+            if (this.PartitionKey != null && request.ResourceType != ResourceType.Document)
+            {
+                throw new ArgumentException($"{nameof(this.PartitionKey)} can only be set for item operations");
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: >ArgumentException [](start = 26, length = 17)

ArgumentOutOfRangeException

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 16]
```diff
@@ -147,9 +149,16 @@ public class QueryRequestOptions : RequestOptions
...
+            }
+
+            // Cross partition is only applicable to item operations.
+            if (this.PartitionKey == null && request.ResourceType == ResourceType.Document)
+            {
+                request.Headers.Add(HttpConstants.HttpHeaders.EnableCrossPartitionQuery, bool.TrueString);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: Why service cares about it? Isn;t it purely for client or max Gateway?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs [hunk 17]
```diff
@@ -436,13 +449,17 @@ internal partial class ContainerCore : Container
                     extractPartitionKeyIfNeeded: true,
                     cancellationToken: cancellationToken);
 
-                ShouldRetryResult retryResult = ShouldRetryResult.NoRetry();
-                if (requestRetryPolicy != null &&
-                    !responseMessage.IsSuccessStatusCode)
+                if (responseMessage.IsSuccessStatusCode)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Nice

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs (hunk 11 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: Reverted

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+                HashSet<string> resultIds = new HashSet<string>(results.Select(x => x.Id));
+                Assert.IsTrue(createdIds.All(x => resultIds.Contains(x)));
+
+                //Basic query
+                List<DatabaseProperties> queryResults = await this.ToListAsync(
+                    client.GetDatabaseQueryStreamIterator,
+                    client.GetDatabaseQueryIterator<DatabaseProperties>,
+                    "select * from T where STARTSWITH(T.id, \"BasicQueryDb\")");
+                resultIds = new HashSet<string>(queryResults.Select(x => x.Id));
+                Assert.IsTrue(createdIds.All(x => resultIds.Contains(x)));
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs]: CollectionAssert is a better alternative.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDatabaseTests.cs [hunk 19]
```diff
@@ -302,8 +303,11 @@ public async Task DatabaseIterator()
                     databaseIds.Add(createResponse.Resource.Id);
                 }
 
-                FeedIterator<DatabaseProperties> feedIterator =
-                    this.cosmosClient.GetDatabaseIterator();
+                FeedIterator<DatabaseProperties> feedIterator = this.cosmosClient.GetDatabaseQueryIterator<DatabaseProperties>(
+                    queryDefinition: null,
+                    continuationToken: null,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDatabaseTests.cs]: Optional

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Utils/TestCommon.cs [hunk 20]
```diff
@@ -1385,7 +1385,11 @@ public static async Task DeleteAllDatabasesAsyncWorker(CosmosClient client)
         {
             IList<Cosmos.Database> databases = new List<Cosmos.Database>();
 
-            FeedIterator<DatabaseProperties> resultSetIterator = client.GetDatabaseIterator(maxItemCount: 10);
+            FeedIterator<DatabaseProperties> resultSetIterator = client.GetDatabaseQueryIterator<DatabaseProperties>(
+                queryDefinition: null,
+                continuationToken: null,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Utils/TestCommon.cs]: optional

File: Microsoft.Azure.Cosmos/src/Resource/QueryResponses/FeedStatelessIteratorCore.cs [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    using System.Runtime.CompilerServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents;
+    using static Microsoft.Azure.Documents.RuntimeConstants;
+
+    /// <summary>
+    /// Cosmos feed stream iterator. This is used to get the query responses with a Stream content
+    /// </summary>
+    internal class FeedStatelessIteratorCore : FeedIterator
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/QueryResponses/FeedStatelessIteratorCore.cs]: >FeedStatelessIteratorCore [](start = 19, length = 25)

Follow-up post PR: Names and hierarchy consolidation.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/QueryResponses/FeedStatelessIteratorCore.cs]: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/489

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDatabaseTests.cs (hunk 19 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDatabaseTests.cs]: I don't see any issue with specifying it. It just makes it clear which contract is getting called.
